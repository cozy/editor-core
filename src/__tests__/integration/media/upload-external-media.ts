import { BrowserTestCase } from '@atlaskit/webdriver-runner/runner';
import {
  editable,
  getDocFromElement,
  fullpage,
  copyToClipboard,
} from '../_helpers';
import {
  goToEditorTestingExample,
  mountEditor,
} from '../../__helpers/testing-example-helpers';
import { sleep } from '@atlaskit/media-test-helpers';
import Page from '@atlaskit/webdriver-runner/wd-wrapper';
import { toBeOneOfMatchers } from './_matchers';
import cloneDeep from 'lodash/cloneDeep';

expect.extend(toBeOneOfMatchers);

const expectFuzzyMediaAttrs = (doc: { [key: string]: any }) => {
  expect(doc.content[0].content[0].attrs).toEqual(
    expect.objectContaining({
      __fileMimeType: expect.stringMatching(/image\/(webp|jpeg)/),
      __fileSize: expect.toBeOneOf([8410, 9894]),
    }),
  );
};

const removeFuzzyMediaAttrs = (doc: { [key: string]: any }) => {
  const copy = cloneDeep(doc);
  delete copy.content[0].content[0].attrs.__fileMimeType;
  delete copy.content[0].content[0].attrs.__fileSize;
  return copy;
};

BrowserTestCase(
  'upload-external-media.ts: Uploads external media when pasted',
  { skip: ['edge', 'safari'] },
  async (client: ConstructorParameters<typeof Page>[0], testCase: string) => {
    const sample = new Page(client);
    await copyToClipboard(
      sample,
      `<meta charset='utf-8'><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/5/5e/Domestic_Cat_Face_Shot.jpg/220px-Domestic_Cat_Face_Shot.jpg"/>`,
      'html',
    );

    const page = await goToEditorTestingExample(client);
    await mountEditor(page, {
      appearance: fullpage.appearance,
      media: {
        allowMediaSingle: true,
      },
    });
    await page.paste();
    await sleep(0);
    //waits until blob is available
    await page.waitForSelector('.ProseMirror img[src^="blob"]');
    let doc = await page.$eval(editable, getDocFromElement);
    // ED-9896: The Accept request header field is modified by
    // by the user-agent when webp support is detected in-browser
    // to proactively set content negotiation preferences. However,
    // we cannot rely on these preferences being consistently honoured
    // by the origin server, so we add separate assertions for these
    // "fuzzier" document results.
    expectFuzzyMediaAttrs(doc);
    expect(removeFuzzyMediaAttrs(doc)).toMatchCustomDocSnapshot(testCase);
  },
);
