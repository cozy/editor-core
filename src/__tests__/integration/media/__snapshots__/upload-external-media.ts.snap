// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload-external-media.ts: Uploads external media when pasted 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "layout": "center",
        "width": null,
      },
      "content": Array [
        Object {
          "attrs": Object {
            "__contextId": "DUMMY-OBJECT-ID",
            "__displayType": null,
            "__external": true,
            "__fileName": "example.png",
            "alt": "",
            "collection": "MediaServicesSample",
            "height": 147,
            "id": "1234-5678-abcd-efgh",
            "occurrenceKey": "12345678-9abc-def0-1234-56789abcdef0",
            "type": "file",
            "url": null,
            "width": 220,
          },
          "type": "media",
        },
      ],
      "type": "mediaSingle",
    },
  ],
  "type": "doc",
}
`;
